lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+            raise UnexpectedResponse(
+                f"The type of the response is {type(result)}, expected: "
+                f"{expected_response_type}"
+            )
+
+        return result
+
+    TSubscribeEvent = TypeVar("TSubscribeEvent", bound=BaseEvent)
+
+    def subscribe(
+        self,
+        event_type: Type[TSubscribeEvent],
+        handler: Callable[[TSubscribeEvent], Union[Any, Awaitable[Any]]],
+    ) -> Subscription:
+        """
+        Subscribe to receive updates for any event that matches the specified event type.
+        A handler is passed as a second argument an :class:`~lahja.common.Subscription` is returned
+        to unsubscribe from the event if needed.
+        """
+        if asyncio.iscoroutinefunction(handler):
+            self._async_handlers[event_type].append(handler)
+            subscription = Subscription(
+                lambda: self._async_handlers[event_type].remove(handler)
+            )
+        else:
+            self._sync_handlers[event_type].append(handler)
+            subscription = Subscription(
+                lambda: self._sync_handlers[event_type].remove(handler)
+            )
+
+        # notify subscriptions have been updated.
+        self._subscriptions_changed.set()
+
+        return subscription
+
+    TStreamEvent = TypeVar("TStreamEvent", bound=BaseEvent)
+
+    async def stream(
+        self, event_type: Type[TStreamEvent], num_events: Optional[int] = None
+    ) -> AsyncGenerator[TStreamEvent, None]:
+        """
+        Stream all events that match the specified event type. This returns an
+        ``AsyncIterable[BaseEvent]`` which can be consumed through an ``async for`` loop.
+        An optional ``num_events`` parameter can be passed to stop streaming after a maximum amount
+        of events was received.
+        """
+        (send_channel, receive_channel) = cast(
+            StreamChannelPair, trio.open_memory_channel(100)
+        )
+
+        self._stream_channels[event_type].add(send_channel)
+
+        # notify subscriptions have been updated.
+        self._subscriptions_changed.set()
+
+        if num_events is None:
+            # iterate forever
+            counter = itertools.count()
+        else:
+            # fixed number of iterations
+            counter = iter(range(max(0, num_events - 1)))
+
+        async for event in receive_channel:
+            yield event  # type: ignore  # mypy doesn't recognize this having a correct type
+
+            try:
+                next(counter)
+            except StopIteration:
+                await send_channel.aclose()
+                break
+
+        # Trigger the subscription change since the event will be removed from
+        # the set once the local reference to the channel goes out of scope.
+        self._subscriptions_changed.set()
+
+    TWaitForEvent = TypeVar("TWaitForEvent", bound=BaseEvent)
+
+    async def wait_for(self, event_type: Type[TWaitForEvent]) -> TWaitForEvent:
```

Comment:
I think this overrides and exactly duplicates `lahja.base.BaseEndpoint.wait_for`